File: Python - pygame/Saucer.py
import pygame, math, random, Global
from Entity import *
from Laser import *
from Explosion import *

class Saucer(Entity):

  def __init__(self, anim):
    super().__init__(anim, 0, Global.HEIGHT * 0.25 + random.randint(0, Global.HEIGHT / 2), 0)
    self.thrustChannel = pygame.mixer.Channel(pygame.mixer.get_num_channels() - 1)
    if Global.saucer is not None:
      self.active = False
      return

    self.dx = random.randint(2, 4) # move at a random speed, mostly horizontal
    self.dy = random.randint(1, 2)
    if (random.randint(0, 1) == 0):
      self.dx *= -1         # randomly move right to left
      self.x = Global.WIDTH # start on the right-hand side of the screen
    if (random.randint(0, 1) == 0):
      self.dy *= -1 # move up or down

  def update(self):
    super().update()
    if not self.thrustChannel.get_busy():
      self.thrustChannel.play(Global.sounds["saucer"])

    if random.randint(0, 150) == 0:
        heading = math.degrees(math.atan2(Global.ship.x - self.x, self.y - Global.ship.y)) - 90
        Laser.spawn(Global.animations["laser_saucer"], self.x, self.y, -heading, True)

    if self.x < 0 or self.x > Global.WIDTH or self.y < 0 or self.y > Global.HEIGHT:
        self.destroy()

  def collides_with(self, other):
    if self.intersects(other):
        self.destroy()
        Explosion.spawn(Global.animations["explosion_ship"], self.x, self.y)
        Global.sounds["bang_lg"].play()
        return True
    return False

  def destroy(self):
    self.thrustChannel.stop()
    self.active = False
    Global.saucer = None

File: Java - AWT - Swing/asteroids/Saucer.java
package asteroids;

public class Saucer extends Entity {

    public Saucer() {
        super(Astro.animations.get("saucer"), 0, Astro.HEIGHT * 0.25f + Astro.random(Astro.HEIGHT / 2), 0);
        if (Astro.saucer != null) {
            active = false; // make sure we only have one saucer at a time
            return;
        }

        dx = 2f + Astro.random(2); // move at a random speed, mostly horizontal
        dy = 1f + Astro.random(1);
        if (Astro.random(1) > 0.5f) {
            dx *= -1; // randomly move right to left
            x = Astro.WIDTH; // start on the right-hand side of the screen
        }
        if (Astro.random(1) > 0.5f)
            dy *= -1; // move up or down
    }

    public void update() {
        super.update();

        if (!Astro.sounds.get("saucer").isActive())
            Astro.play(Astro.sounds.get("saucer"));

        if ((int) Astro.random(150) == 0) {
            float heading = (float) (Math.atan2(Astro.ship.x - x, y - Astro.ship.y) - Math.PI * 0.5);
            Laser.spawn(x, y, heading, true);
        }
        if (x < 0 || x > Astro.WIDTH || y < 0 || y > Astro.HEIGHT)
            destroy();
    }

    public boolean collides_with(Entity e) {
        if (intersects(e)) {
            destroy();
            Explosion.spawn(Astro.animations.get("explosion_ship"), x, y);
            Astro.play(Astro.sounds.get("bang_lg"));
            return true;
        }
        return false;
    }

    public void destroy() {
        Astro.saucer = null;
        active = false;
    }
}
